Add plumbing for mod-defined sound loaders

penev92 opened the pull request:
Use cases are #10032 and moving `AudLoader` and `WavLoader` out of the engine and into Common.
This brings sound loaders on par with sprite loaders and their `ISpriteLoader` and enables every mod to define it's own loaders.

Note: this is for sound effects/voices only. Music is still [hardcoded](https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Game/GameRules/MusicInfo.cs#L43-L48). Fixing that both is out of scope here and will conflict with the unglobaling of the file system.

penev92 committed:
Add ISoundLoader

The interface is to be implemented by all sound loaders, just like ISpriteLoader. All loading goes through the interface. This would allow mods to create their own sound loaders outside the engine.
Also add a SoundFormats property to mod.yaml, where mods can define what sound loaders they will need.
This requires Game.Sound to be initialized after the ModData is loaded.

penev92 committed:
Make WavLoader implement ISoundLoader

penev92 committed:
Make AudLoader implement ISoundLoader

penev92 committed:
Add the SoundFormats field to mods

abcdefg30 commented:
> OpenRA.Game/FileFormats/AudLoader.cs:L49: [SA1311] Static readonly fields must start with an upper-case letter: indexAdjust.
OpenRA.Game/FileFormats/AudLoader.cs:L50: [SA1311] Static readonly fields must start with an upper-case letter: stepTable.

penev92 commented:
Fixed.

penev92 commented:
As discussed today on IRC this should be the first PR of the two to be merged so the other can adapt to the interface.

RoosterDragon commented on OpenRA.Game/ModData.cs:
```diff
@@ -53,6 +54,18 @@ public ModData(string mod, bool useLoadScreen = false)
+				soundLoaders.Add((ISoundLoader)ObjectCreator.CreateBasic(loader));
+			}
+
+			SoundLoaders = soundLoaders.ToArray();
```
Feels like you could extract a common method for the sound and sprite loaders.

RoosterDragon commented:
May be worth considering adding a try-catch around the internals of the sound parsing so you can return false for `TryParseSound` even when an exception occurs.

penev92 commented:
The internal parsing code was intentionally left volatile so people couldn't get away with bogus files (or get the less helpful "No suitable parsers found" exception).

RoosterDragon commented:
You can still throw in the normal case.

penev92 committed:
Add ModData.GetLoaders()

Extract sound loaders and sprite loaders creation into a generic method to avoid code duplication.

penev92 committed:
Don't crash when parsing corrupt sound files

Log the exception and move to the next parser.

penev92 commented:
Added two new commits with the requested changes.

GraionDilach commented:
That last commit sounds like a workaround for #10124.

pchote commented on OpenRA.Game/FileFormats/AudLoader.cs:
```diff
@@ -153,7 +166,33 @@ public static byte[] LoadSound(Stream s)
+			}
+			catch (Exception e)
+			{
+				Log.Write("debug", "Failed to parse AUD file {0}. Error message:".F(fileName));
```
You're using TryParseSound for format detection as well as just loading.  This means you will spam the debug logs with "failed to load aud" errors for each valid wav if you define aud as the first loader in the list.

pchote replied:
If you want to keep the current interface/api you will need to keep the try/catch, but remove the logging.

penev92 replied:
Actually no.
The inner parsing method (`LoadSound()` in this case) returns `true`/`false` depending on whether the file is in a format this parser accepts. So the `AudLoader` will just return `false` for WAV files and vice-versa instead of throwing.
The exceptions happen when the file is corrupt or otherwise bogus - ends prematurely, the metadata at the start doesn't correspond to the actual data (like in #10124) or other such situations. So only actual exceptions will be logged.

Perhaps I should add a comment to make that clear by just glancing at the code.

pchote replied:
Ok, makes sense.  Yes, please do add a comment.

penev92 commented:
Added the comments.

abcdefg30 commented:
Needs a rebase.

penev92 commented:
Rebased.

RoosterDragon commented:
:+1: Looks alright to me and sounds still work.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/40c248d9126f89c303cbf2696d4274f60df7984a...bd1daadb65b51ca711b3b3f2db2c790f4893cbd3)
